starlite/utils/sync.py

```diff
@@ -0,0 +1,29 @@
+from functools import partial
+from typing import Callable, TypeVar
+
+from anyio.to_thread import run_sync
+from typing_extensions import ParamSpec
+
+from starlite.utils.predicates import is_async_callable
+
+P = ParamSpec("P")
+T = TypeVar("T")
+
+
+class AsyncCallable:
+    __slots__ = ("args", "kwargs", "fn")
+
+    def __init__(self, fn: Callable[P, T]):
+        """Utility class that wraps a callable and ensures it can be called as
+        an async function.
+
+        Args:
+            fn: Callable to wrap - can be any sync or async callable.
+        """
+        if is_async_callable(fn):
+            self.fn = fn
+        else:
+            self.fn = partial(run_sync, fn)  # type: ignore
+
+    async def __call__(self, *args: P.args, **kwargs: P.kwargs) -> T:
+        return await self.fn(*args, **kwargs)  # type: ignore
```

Comment:
I remember looking at paramspec for `LifecycleHook` but I couldn't get it to behave the way I thought it should. 

We are trying to get this type of thing to type correctly right?

```python
from starlite.utils.sync import AsyncCallable


def fn(y: int, z: str) -> str:
    ...


async def test() -> None:
    a = AsyncCallable(fn=fn)
    await a(1, "two")
```
Mypy:
```
scratch_8.py:10: error: Argument 1 to "__call__" of "AsyncCallable" has incompatible type "int"; expected <nothing>  [arg-type]
scratch_8.py:10: error: Argument 2 to "__call__" of "AsyncCallable" has incompatible type "str"; expected <nothing>  [arg-type]
```

One thing that stands out is that without binding `P` to the class the reference to `P` in signature for both `__init__` and `__call__` probably doesn't correlate with eachother.

A couple of quick examples I've found using paramspec and `__call__()` both make the class generic on `P` before referencing in signatures, e.g., https://github.com/python/mypy/issues/12595 and https://github.com/python/mypy/issues/12169.